scripts/main.js

```diff
@@ -78,462 +78,459 @@ let keysObj2 = {
+    if (keyboardArr2 !== null) {
+      keysContainer2.classList.add("keysContainer2");
+      keysContainer2.style.display = defDsply2;
+      keysContainer2.style.flexFlow = "row wrap";
+      keysContainer2.style.justifyContent = "center";
+      keysContainer2.style.alignItems = "center";
+      keyboardLayout.appendChild(keysContainer2);
+    }
+
+    // SPECIAL BUTTONS INJECTOR
+    let btnInjector = (array) => {
+
+      // button objects with special functions
+      const capsLock = {
+        value: "",
+        style: {
+          width: "13%"
+        },
+        class: {
+          0: "fal",
+          1: "fa-arrow-alt-up",
+        },
+        operator: () => {
+          let btn = document.querySelectorAll(".keysContainer1 button");
+          let indicatorBtn = document.querySelector(".fa-arrow-alt-up div");
+
+          for (let i = 0; i < btn.length; i++) {
+
+            if (/[a-zA-Z]/.test(btn[i].innerText) && btn[i].innerText != switcher2) {
+              let isLowerCase = true;
+
+              if (/[a-z]/.test(btn[i].innerHTML)) {
+                btn[i].innerText = btn[i].innerText.toUpperCase();
+                indicatorBtn.style.background = " rgb(16, 230, 16)";
+                isLowerCase = false;
+              }
+
+              if (isLowerCase) {
+                indicatorBtn.style.background = "rgba(218, 218, 218, 0.6)";
+                btn[i].innerText = btn[i].innerText.toLowerCase();
+              }
+            }
+          }
+        }
+      }
+
+      const backSpace = {
+        value: "",
+        style: {
+          width: "13%"
+        },
+        class: {
+          0: "fal",
+          1: "fa-backspace"
+        },
+        operator: () => {
+          console.clear();
+
+          // if caret already at 0 index, do NOTHING!
+          if (input.selectionStart === 0) return;
+
+          // get access to the input value
+          let txt = input.value;
+
+          // set start index conditionally:
+          //  one position behind the caret if no selection is made, or
+          //  no modification, if a selection range is made
+          let startIdx =
+            input.selectionStart === input.selectionEnd
+              ? input.selectionStart - 1
+              : input.selectionStart;
+
+          let endIdx = input.selectionEnd;
+
+          // declare a "pre-selection" slice of text, a "middle" which will be the deleted portion, and a "post-selection" slice of text
+          let pre = txt.slice(0, startIdx);
+          let mid = txt.slice(startIdx, endIdx);
+          let post = txt.slice(endIdx);
+
+          // concat pre and post, leaving out the selected middle
+          let newTxt = pre.concat(post);
+
+          // update the input with the new chars
+          input.value = newTxt;
+
+          // set caret with offset to account for deleted characters
+          input.setSelectionRange(startIdx, startIdx);
+
+          // not sure what this does
```

Comment:
The best way to explain this is that you experience it yourself. First, fire up the original code in a live server, and comment this out, like you just did here. Then type something and position the caret at the index 0 and hit backspace. 

The caret will jump to the end of the line (not sure why)! This code prevents it from happening. My initial solution was to use the same code you did above:

if (input.selectionStart === 0) return;

However, you already know what that'll cause, as I commented. So I came up with this temporary syntax which I think it might make the range deletion starting or ending at index 0, possible. I have not come up with final solution yet.

Reply:
Ahh ok.  This seems fine - since it is at the end of the script, it won't interrupt any legal delete actions (like a range that starts at 0).  It will just prevent the skipping behavior - nice.  I removed my "early return" in the first conversation and uncommented this one.  Not sure if any final improvement exists...seems fine!